### PR TITLE
Page list filter

### LINKED ISSUE #240: Pages list filter
Add filter options to the pages overview. The following checkboxes should be available:

- [ ] "Translation is up-to-date" / "Übersetzung ist aktuell"
- [ ] "Translation is outdated" / "Übersetzung ist veraltet"
- [ ] "Translation is missing" / "Übersetzung fehlt"

Also, drag and drop should be deactivated when filters are applied.

### PR DESCRIPTION
**Pull request by Thomas125**
### Short description
Add filtering to the page tree

### Proposed changes
- Adapt the filtering solution of events also for pages

### Resolved issues
Resolves #240

### CODE REVIEW

File: src/cms/templates/pages/page_tree_node.html [hunk 1]
```diff
@@ -4,11 +4,13 @@
...
-            <i data-feather="move"></i>
-        </span>
-    </td>
+    {% if enable_drag_and_drop %}
+        <td class="single_icon">
+            <span data-drag-id="{{ page.id }}" data-node-descendants="{{ page|get_descendants }}" class="drag text-gray-800 block py-3 pl-4 pr-2 cursor-move" draggable="true">
+                <i data-feather="move"></i>
+            </span>
+        </td>
+    {% endif %}
```
Comment (timoludwig) [inline src/cms/templates/pages/page_tree_node.html:13]:
```suggestion
    <td class="single_icon">
        <span data-drag-id="{{ page.id }}"
              data-node-descendants="{{ page|get_descendants }}"
              class="{% if enable_drag_and_drop %}drag cursor-move{% else %}cursor-not-allowed{% endif %} text-gray-800 block py-3 pl-4 pr-2"
              {% if enable_drag_and_drop %}
              draggable="true"
              {% else %}
              title="{% trans 'Drag & drop is disabled when filters are applied.' %}"
              {% endif %}>
            <i data-feather="move"></i>
        </span>
    </td>
    {% endif %}
```
I would propose to show the hierarchy even if drag & drop is disabled. Also, the checkboxes look a but buggy if they're not aligned...

File: src/cms/templates/pages/page_tree.html [hunk 2]
```diff
@@ -47,16 +52,19 @@ <h2 class="heading font-normal mb-2">{% trans 'Pages' %}</h2>
...
+<div id="filter-form-container" class="{% if not filter_form.has_changed %}hidden{% endif %} w-full mt-4 rounded border border-solid border-gray-200 shadow bg-white">
+    {% include 'pages/_page_filter_form.html' %}
+</div>
 
 <div class="table-listing">
     <table class="w-full mt-4 rounded border border-solid border-gray-200 shadow bg-white table-auto">
         <thead>
             <tr class="border-b border-solid border-gray-200">
-                <th class="text-sm text-left uppercase py-3 pl-4 pr-2 min"></th>
+                {% if enable_drag_and_drop %}<th class="text-sm text-left uppercase py-3 pl-4 pr-2 min"></th>{% endif %}
```
Comment (timoludwig) [inline src/cms/templates/pages/page_tree.html:67]:
```suggestion
                <th class="text-sm text-left uppercase py-3 pl-4 pr-2 min"></th>
```
If you apply my previous suggestion, this table column is required.

File: src/cms/views/pages/page_tree_view.py [hunk 3]
```diff
@@ -4,23 +4,17 @@
...
 @method_decorator(login_required, name="dispatch")
 @method_decorator(region_permission_required, name="dispatch")
-# pylint: disable=too-many-ancestors
-class PageTreeView(PermissionRequiredMixin, TemplateView, PageMixin):
-    """
-    View for showing the page tree
-    """
-
-    #: Required permission of this view (see :class:`~django.contrib.auth.mixins.PermissionRequiredMixin`)
+class PageTreeView(PermissionRequiredMixin, TemplateView):
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:17]:
```suggestion
# pylint: disable=too-many-ancestors
class PageTreeView(PermissionRequiredMixin, TemplateView, PageMixin):
    """
    View for showing the page tree
    """

    #: Required permission of this view (see :class:`~django.contrib.auth.mixins.PermissionRequiredMixin`)
```
Please do not remove code without asking :wink: 
The `PageMixin` is required for the confirmation popups when archiving or deleting pages from the page tree.

File: src/cms/views/pages/page_tree_view.py [hunk 4]
```diff
@@ -4,23 +4,17 @@
 from django.shortcuts import render, redirect
 from django.utils.decorators import method_decorator
 from django.utils.translation import ugettext as _
-
 from django.views.generic import TemplateView
 
+from ...constants import translation_status
 from ...decorators import region_permission_required
+from ...forms.pages.page_filter_form import PageFilterForm
 from ...models import Region, Language
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:12]:
```suggestion
from ...models import Region, Language
from .page_mixin import PageMixin
```
The `PageMixin` is required.

File: src/cms/views/pages/page_tree_view.py [hunk 5]
```diff
@@ -93,15 +88,60 @@ def get(self, request, *args, **kwargs):
                 request, _("You don't have the permission to edit or create pages.")
             )
         context = self.get_context_data(**kwargs)
+
+        pages = region.get_pages(archived=self.archived)
+        enable_drag_and_drop = True
+        # Filter pages according to given filters, if any
+        filter_data = kwargs.get("filter_data")
+        if filter_data is not None:
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:96]:
```suggestion
        if filter_data:
```
This is enough

File: src/cms/views/pages/page_tree_view.py [hunk 6]
```diff
@@ -93,15 +88,60 @@ def get(self, request, *args, **kwargs):
...
+            filter_form = PageFilterForm(data=filter_data)
+            if filter_form.is_valid():
+                # only filter if at least one checkbox but not all are checked
+                if (
+                    0
+                    < len(filter_form.cleaned_data["translation_status"])
+                    < len(translation_status.CHOICES)
+                ):
+                    enable_drag_and_drop = False
+                    utd = (
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:107]:
```suggestion
                    filter_up_to_date = (
```
I prefer speaking variable names. Even if it's not very ambiguous in this case, there is no reason to be stingy with letters :wink:

File: src/cms/views/pages/page_tree_view.py [hunk 7]
```diff
@@ -93,15 +88,60 @@ def get(self, request, *args, **kwargs):
...
+                        translation_status.UP_TO_DATE
+                        in filter_form.cleaned_data["translation_status"]
+                    )
+                    m = (
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:111]:
```suggestion
                    filter_missing = (
```

File: src/cms/views/pages/page_tree_view.py [hunk 8]
```diff
@@ -93,15 +88,60 @@ def get(self, request, *args, **kwargs):
...
+                        translation_status.MISSING
+                        in filter_form.cleaned_data["translation_status"]
+                    )
+                    od = (
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:115]:
```suggestion
                    filter_outdated = (
```

File: src/cms/views/pages/page_tree_view.py [hunk 9]
```diff
@@ -93,15 +88,60 @@ def get(self, request, *args, **kwargs):
...
+                        translation_status.OUTDATED
+                        in filter_form.cleaned_data["translation_status"]
+                    )
+
+                    def f(page):
+                        translation = page.get_translation(language_code)
+                        if translation is None:
+                            return m
+                        return od if translation.is_outdated else utd
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:124]:
```suggestion
                    def page_filter(page):
                        translation = page.get_translation(language_code)
                        if not translation:
                            return filter_missing
                        elif translation.is_outdated
                            return filter_outdated
                        return filter_up_to_date
```
I find this slightly more readable.

File: src/cms/views/pages/page_tree_view.py [hunk 10]
```diff
@@ -93,15 +88,60 @@ def get(self, request, *args, **kwargs):
...
+
+                    pages = list(filter(f, pages))
```
Comment (timoludwig) [inline src/cms/views/pages/page_tree_view.py:126]:
```suggestion
                    pages = list(filter(page_filter, pages))
```

**Review by timoludwig — COMMENTED** [review]
Thanks! In general, the feature works as expected :+1: 
I just have a few comments on the implementation details.

File: src/cms/views/pages/page_tree_view.py (hunk 3 above)
Comment (Thomas125) [reply src/cms/views/pages/page_tree_view.py]: Mhm, this wasn't on purpose. I can't remember changing this. Now, I changed it back

File: src/cms/templates/pages/_page_filter_form.html [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    </div>
+</form>
+
+{% block javascript %}
+    <script src={% static 'js/reset_filter.js' %}></script>
+    <script>
+        // event handler to reset filter form
+        u('#filter-reset').on('click', (event) => {
+            reset_filter(event);
+        });
```
Comment (timoludwig) [inline src/cms/templates/pages/_page_filter_form.html:31]:
```suggestion
        u('#filter-reset').on('click', reset_filter);
```
You don't need an anonymous function just to call another function, you can directly use the callback as an argument here.

**Review by timoludwig — COMMENTED** [review]
Thanks for your changes :+1: 
I have one tiny idea for improvement, but in general this PR looks good! :rocket:

**Review by timoludwig — APPROVED** [review]
Thanks! You can now merge this PR :tada:

**Pull request merged**